beacon-chain/sync/initial-sync/blocks_fetcher_test.go

```diff
@@ -0,0 +1,362 @@
+					finalizedEpoch: 18,
+					headSlot:       640,
+				},
+				{
+					blocks:         append(makeSequence(1, 64), makeSequence(500, 640)...),
+					finalizedEpoch: 18,
+					headSlot:       640,
+				},
+			},
+			requests: []*fetchRequestParams{
+				{
+					start: 1,
+					count: blockBatchSize,
+				},
+				{
+					start: blockBatchSize + 1,
+					count: blockBatchSize,
+				},
+				{
+					start: 2*blockBatchSize + 1,
+					count: blockBatchSize,
+				},
+				{
+					start: 400,
+					count: 150, // [400, 550), with blocks availability starting at 500. So, 500, 501, ..., 550
+				},
+				{
+					start: 551,
+					count: 200,
+				},
+			},
+		},
+		{
+			name:               "Multiple peers with failures",
+			expectedBlockSlots: makeSequence(1, 2*blockBatchSize),
+			peers: []*peerData{
+				{
+					blocks:         makeSequence(1, 320),
+					finalizedEpoch: 8,
+					headSlot:       320,
+				},
+				{
+					blocks:         makeSequence(1, 320),
+					finalizedEpoch: 8,
+					headSlot:       320,
+					failureSlots:   makeSequence(1, 32), // first epoch
+				},
+				{
+					blocks:         makeSequence(1, 320),
+					finalizedEpoch: 8,
+					headSlot:       320,
+				},
+				{
+					blocks:         makeSequence(1, 320),
+					finalizedEpoch: 8,
+					headSlot:       320,
+				},
+			},
+			requests: []*fetchRequestParams{
+				{
+					start: 1,
+					count: blockBatchSize,
+				},
+				{
+					start: blockBatchSize + 1,
+					count: blockBatchSize,
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			initializeRootCache(tt.expectedBlockSlots, t)
+
+			beaconDB := dbtest.SetupDB(t)
+
+			p := p2pt.NewTestP2P(t)
+			connectPeers(t, p, tt.peers, p.Peers())
+			genesisRoot := rootCache[0]
+
+			err := beaconDB.SaveBlock(context.Background(), &eth.SignedBeaconBlock{
+				Block: &eth.BeaconBlock{
+					Slot: 0,
+				}})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			st, err := stateTrie.InitializeFromProto(&p2ppb.BeaconState{})
+			if err != nil {
+				t.Fatal(err)
+			}
+			mc := &mock.ChainService{
+				State: st,
+				Root:  genesisRoot[:],
+				DB:    beaconDB,
+			} // no-op mock
+
+			ctx, cancel := context.WithCancel(context.Background())
+			fetcher := newBlocksFetcher(&blocksFetcherConfig{
+				ctx:         ctx,
+				chain:       mc,
+				p2p:         p,
+				rateLimiter: leakybucket.NewCollector(allowedBlocksPerSecond, allowedBlocksPerSecond, false /* deleteEmptyBuckets */),
+			})
+
+			fetcher.start()
+
+			var wg sync.WaitGroup
+			wg.Add(len(tt.requests)) // how many block requests we are going to make
+			go func() {
+				wg.Wait()
+				log.Debug("Stopping fetcher")
+				fetcher.stop()
+			}()
+
+			processFetchedBlocks := func() ([]*eth.SignedBeaconBlock, error) {
+				defer cancel()
+				var unionRespBlocks []*eth.SignedBeaconBlock
+
+				for {
+					select {
+					case <-time.After(5 * time.Second): // TODO(4815): Temporary safeguard, remove
```

Comment:
Agreed! The only reason it is structured like that is that, I wanted to make it as close to round robin tests as possible, as I planned to swap the current implementation of round robing with my refactored code - so wanted to increase chances it will work w/o issues. Still, we can refactor further, and split the code into more manageable chunks (for both using and tests).